### PR TITLE
Add pre-condition check for fuse optimizer op pass

### PR DESCRIPTION
**Pull request by chenwhql**
gradients of optimizer op to a contiguous address on the GPU device **before the main graph is executed**, since some grad operators may not have a GPU kernel, the gradients gerarated cannot be written to the GPU device as expected, but generated separately on the CPU device, which will cause the program to use dirty data in GPU device to continue training, it may cause calculate error or NAN. 

Related PR：https://github.com/PaddlePaddle/Paddle/pull/19966

This PR add two pre-condition check, if pre-condition is not met, this pass will not be applyed.
1. The fusing gradients generated op does not support GPU device.
2. The dtypes of fusing variables are different. (originally this case will throw exception directly)

Other change:
1. add else branch for output date init in coalesce tensor op.
2. polish code & add error messages.

I construct a test case by deleting the CUDA kernel of mul_op, so if set `fuse_all_optimizer_ops = True` in a simple fc net, it will be ignored when PE is executed. The test log is as follows.
```
I1107 06:32:30.669647 11004 fuse_optimizer_op_pass.cc:55] Find momentum operators : 2, and 2 for dense gradients.
I1107 06:32:30.669725 11004 fuse_optimizer_op_pass.cc:96] Velocity: @FUSEDVAR@_momentum_Velocity_fc_0.b_0_velocity_0
I1107 06:32:30.669760 11004 fuse_optimizer_op_pass.cc:96] Param: @FUSEDVAR@_momentum_Param_fc_0.b_0
I1107 06:32:30.669776 11004 fuse_optimizer_op_pass.cc:96] Grad: @FUSEDVAR@_momentum_Grad_fc_0.b_0@GRAD
I1107 06:32:30.669807 11004 fuse_optimizer_op_pass.cc:450] Grad: fc_0.b_0@GRAD fc_0.w_0@GRAD
I1107 06:32:30.669836 11004 fuse_optimizer_op_pass.cc:450] Param: fc_0.b_0 fc_0.w_0
I1107 06:32:30.669860 11004 fuse_optimizer_op_pass.cc:450] Velocity: fc_0.b_0_velocity_0 fc_0.w_0_velocity_0
I1107 06:32:30.669883 11004 fuse_optimizer_op_pass.cc:271] Op check: elementwise_add_grad, support CPU: 1, support GPU: 1
I1107 06:32:30.669915 11004 fuse_optimizer_op_pass.cc:271] Op check: mul_grad, support CPU: 1, support GPU: 0
```

another example:
```
   680 272: I1107 06:43:18.733654 30643 fuse_optimizer_op_pass.cc:55] Find sgd operators : 10, and 10 for dense gradients.
   681 272: I1107 06:43:18.733804 30643 fuse_optimizer_op_pass.cc:96] Param: @FUSEDVAR@_sgd_Param_batch_norm_0.b_0
   682 272: I1107 06:43:18.733824 30643 fuse_optimizer_op_pass.cc:96] Grad: @FUSEDVAR@_sgd_Grad_batch_norm_0.b_0@GRAD
   683 272: I1107 06:43:18.733870 30643 fuse_optimizer_op_pass.cc:271] Op check: batch_norm_grad, support CPU: 1, support GPU: 1
   684 272: I1107 06:43:18.733889 30643 fuse_optimizer_op_pass.cc:271] Op check: sum, support CPU: 1, support GPU: 1
   685 272: I1107 06:43:18.733899 30643 fuse_optimizer_op_pass.cc:271] Op check: batch_norm_grad, support CPU: 1, support GPU: 1
   686 272: I1107 06:43:18.733904 30643 fuse_optimizer_op_pass.cc:271] Op check: sum, support CPU: 1, support GPU: 1
   687 272: I1107 06:43:18.733911 30643 fuse_optimizer_op_pass.cc:271] Op check: batch_norm_grad, support CPU: 1, support GPU: 1
   688 272: I1107 06:43:18.733916 30643 fuse_optimizer_op_pass.cc:271] Op check: sum, support CPU: 1, support GPU: 1
   689 272: I1107 06:43:18.733922 30643 fuse_optimizer_op_pass.cc:271] Op check: batch_norm_grad, support CPU: 1, support GPU: 1
   690 272: I1107 06:43:18.733927 30643 fuse_optimizer_op_pass.cc:271] Op check: sum, support CPU: 1, support GPU: 1
   691 272: I1107 06:43:18.733932 30643 fuse_optimizer_op_pass.cc:271] Op check: sum, support CPU: 1, support GPU: 1
   692 272: I1107 06:43:18.733939 30643 fuse_optimizer_op_pass.cc:271] Op check: elementwise_add_grad, support CPU: 1, support GPU: 1
   693 272: I1107 06:43:18.733947 30643 fuse_optimizer_op_pass.cc:271] Op check: mul_grad, support CPU: 1, support GPU: 0
```

### CODE REVIEW

File: paddle/fluid/operators/coalesce_tensor_op.cc [hunk 1]
```diff
@@ -92,6 +107,8 @@ class CoalesceTensorOp : public framework::OpKernel<T> {
       math::SetConstant<DeviceContext, T> set_constant;
       set_constant(dev_ctx, fused_tensor,
                    static_cast<T>(context.Attr<float>("constant")));
+    } else {
+      math::set_constant(dev_ctx, fused_tensor, 0.0);
```
Comment (chenwhql) [inline paddle/fluid/operators/coalesce_tensor_op.cc]: now in other places, there are usages where `copy_data` and `set_constant` are both false, so here init to zero.

File: paddle/fluid/framework/ir/fuse_optimizer_ops_pass/fuse_optimizer_op_pass.cc [hunk 2]
```diff
@@ -137,54 +136,77 @@ void FuseOptimizerOpPass::ApplyImpl(ir::Graph *graph) const {
...
+        //   in mixed precision scenarios is not yet supported.
+        return;
+      }
+    }
+  }
+
+  // Pass pre-condition check: gradients generated op kernel
+  auto fusing_grad_var_names = aux_var_map.at(kGrad);
+  for (auto grad_var_name : fusing_grad_var_names) {
+    if (!GradGeneratedOpSupportGPU(vars_info, grad_var_name)) {
```
Comment (sneaxiy) [inline paddle/fluid/framework/ir/fuse_optimizer_ops_pass/fuse_optimizer_op_pass.cc]: Seems that it makes this pass does not support CPU case.

File: paddle/fluid/operators/coalesce_tensor_op.cc (hunk 1 above)
Comment (sneaxiy) [inline paddle/fluid/operators/coalesce_tensor_op.cc]: Why call `set_constant`?

Comment (chenwhql) [reply paddle/fluid/operators/coalesce_tensor_op.cc]: done, remove it, now no else branch is no error.

File: paddle/fluid/framework/ir/fuse_optimizer_ops_pass/fuse_optimizer_op_pass.cc (hunk 2 above)
Comment (chenwhql) [reply paddle/fluid/framework/ir/fuse_optimizer_ops_pass/fuse_optimizer_op_pass.cc]: done, check whether op with kernel both support CPU and GPU

**Review by sneaxiy — APPROVED** [review]
LGTM. Waiting for your next PR to refine this strategy again.

**Pull request merged**